changelog/@unreleased/pr-4715.v2.yml

```diff
@@ -0,0 +1,9 @@
+type: improvement
+improvement:
+  description: |-
+    PaxosStateLogImpl and PaxosLeaderElectionService use structured
+    locks (the synchronized keyword) for better observability.
+    This makes it easier to discover which thread is holding locks
+    while others wait from a jstack.
```

Comment:
Why is this not handled correctly by the JVM? Feels like a pretty big gaping hole if the locking classes hinder observability

Reply:
I meant that jstack can point out which thread holds a given lock, and which are waiting for it.

Reply:
I don’t think we should ever be using external synchronisation

Reply:
We shouldn't pass null values, but we still verify using Preconditions.checkNotNull ;-)
The tricky bit is when third party code or API consumers end up synchronizing on your objects.

Reply:
I don't think we run checkNotNull except on external API. This should be internal API, unless I'm mistaken.